File: 0493-reverse-pairs/0493-reverse-pairs.py
class Solution(object):
    def reversePairs(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        #For merging the two array based on index
        def merge(nums, low, mid, high):
            left = low
            right = mid + 1
            li = []

            while left <= mid and right <= high:
                if nums[left] <= nums[right]:
                    li.append(nums[left])
                    left += 1
                else:
                    li.append(nums[right])
                    right += 1

            while left <= mid:
                li.append(nums[left])
                left += 1

            while right <= high:
                li.append(nums[right])
                right += 1

            nums[low:high + 1] = li

            
        
        # Counting the Valid pair 
        def countThePair(num,low,mid,high) :
            count = 0
            
            ptr = mid + 1
            for i in range(low,mid+1) :
                while ptr <= high and nums[i] > 2*nums[ptr] : ptr += 1
                count += (ptr - (mid + 1))
            return count
            
            
        def getPairCount(nums,low,high) :
            count = 0
            if low < high :
                mid = (low + high) // 2
                count += getPairCount(nums,low,mid)
                count += getPairCount(nums,mid+1,high)
                count += countThePair(nums,low,mid,high)
                merge(nums,low,mid,high)
            
            return count
        
        
        
        
        
        return getPairCount(nums,0,len(nums)-1)
        

File: 0493-reverse-pairs/0493-reverse-pairs.java
class Solution {
    public int merge(int[] nums,int low,int mid,int high){
        int[] temp = new int[high - low + 1];
        int count = 0 , left = low , right = mid + 1 , index = 0;
        int leftPointer = low , rightPointer = mid + 1;
        
        while(leftPointer <= mid && rightPointer <= high){
            if((long)nums[leftPointer] > (long)2*nums[rightPointer]){
                count += (mid - leftPointer + 1);
                rightPointer++;
            }
            else{
                leftPointer++;
            }
        }
        
        while(left <= mid && right <= high){
            if(nums[left] > nums[right]){
                temp[index++] = nums[right++];
            }
            else{
                temp[index++] = nums[left++];
            }
        }
        
        while(left <= mid ){
                temp[index++] = nums[left++];   
        }
        
        while(right <= high){
                temp[index++] = nums[right++];
        }
        
        for(int i=0;i<high-low+1;i++){
            nums[low+i] = temp[i];
        }
        return count;
    }
    

    public int mergeSort(int[] nums,int low,int high){
        int count = 0;
        if(low >= high) return 0;
        int mid = (low+high) / 2;
        
            
        count += mergeSort(nums,low,mid) + mergeSort(nums,mid+1,high);
            
        count += merge(nums,low,mid,high);

        return count;
    }
    
    
    public int reversePairs(int[] nums) {
        return mergeSort(nums,0,nums.length-1);
    }
}